private/buf/cmd/buf/command/alpha/repo/reposync/reposync.go

```diff
@@ -0,0 +1,473 @@
+	}
+	return sync(
+		ctx,
+		container,
+		flags.Modules,
+		// No need to pass `flags.Create`, this is not empty iff `flags.Create`
+		flags.CreateVisibility,
+	)
+}
+
+func sync(
+	ctx context.Context,
+	container appflag.Container,
+	modules []string,
+	createWithVisibility string,
+) error {
+	// Assume that this command is run from the repository root. If not, `OpenRepository` will return
+	// a dir not found error.
+	repo, err := git.OpenRepository(git.DotGitDir, command.NewRunner())
+	if err != nil {
+		return err
+	}
+	defer repo.Close()
+	storageProvider := storagegit.NewProvider(
+		repo.Objects(),
+		storagegit.ProviderWithSymlinks(),
+	)
+	knownTags := map[string][]string{}
+	if err := repo.ForEachTag(func(tag string, commitHash git.Hash) error {
+		knownTags[commitHash.Hex()] = append(knownTags[commitHash.Hex()], tag)
+		return nil
+	}); err != nil {
+		return fmt.Errorf("range tags: %w", err)
+	}
+	clientConfig, err := bufcli.NewConnectClientConfig(container)
+	if err != nil {
+		return fmt.Errorf("create connect client %w", err)
+	}
+	var syncableModules []syncableModule
+	for _, module := range modules {
+		var moduleIdentityOverride bufmoduleref.ModuleIdentity
+		if colon := strings.IndexRune(module, ':'); colon != -1 {
+			moduleIdentityOverride, err = bufmoduleref.ModuleIdentityForString(module[colon+1:])
+			if err != nil {
+				return err
+			}
+			module = normalpath.Normalize(module[:colon])
+		}
+		syncableModules = append(syncableModules, syncableModule{
+			dir:      module,
+			identity: moduleIdentityOverride,
+		})
+	}
+	for _, branch := range []string{repo.BaseBranch()} {
+		// TODO: sync other branches
+		if err := syncBranch(
+			ctx,
+			container,
+			clientConfig,
+			repo,
+			branch,
+			storageProvider,
+			knownTags,
+			syncableModules,
+			createWithVisibility,
+		); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func syncBranch(
+	ctx context.Context,
+	container appflag.Container,
+	clientConfig *connectclient.Config,
+	repo git.Repository,
+	branch string,
+	storagegitProvider storagegit.Provider,
+	knownTags map[string][]string,
+	modules []syncableModule,
+	createWithVisibility string,
+) error {
+	// TODO: resume from last sync point
+	return repo.ForEachCommit(branch, func(commit git.Commit) error {
+		for _, module := range modules {
+			if err := pushModuleCommit(
+				ctx,
+				container,
+				clientConfig,
+				module.dir,
+				module.identity,
+				repo,
+				branch,
+				commit,
+				storagegitProvider,
+				knownTags,
+				createWithVisibility,
+			); err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+func pushModuleCommit(
+	ctx context.Context,
+	container appflag.Container,
+	clientConfig *connectclient.Config,
+	moduleDir string,
+	moduleIdentityOverride bufmoduleref.ModuleIdentity,
+	repo git.Repository,
+	branch string,
+	commit git.Commit,
+	storagegitProvider storagegit.Provider,
+	knownTags map[string][]string,
```

Comment:
iiuc this func is pushing a single commit, so you could receive only those tags here
```suggestion
	commit git.Commit,
	storagegitProvider storagegit.Provider,
	commitTags []string,
```